### ISSUE TITLE
pgbouncer crashes when user is equal to auth_user in database config

### ISSUE THREAD

**Issue by sidot3291** [pgbouncer/pgbouncer#314]
For example:
```
test_database =  host=127.0.0.1 port=5432 auth_user=test_user user=test_user password=password
```

- If you attempt to login with a user defined in auth_file, the connection will succeed.
- If you attempt to login with a user not defined in auth_file and not returned by auth_query, you will receive a no such user error.
- If you attempt to login with a user not defined in auth_file but is returned by auth_query, pgbouncer crashes with:

> FATAL @src/objects.c:310 in function put_in_order(): put_in_order: found existing elem

This is true for both 'md5' and 'trust' auth types.

**Comment by petere** [comment]
I cannot reproduce this. Can you send your complete setup, `pgbouncer.ini`, `userlist.txt`, `pg_authid` contents, etc.?

**Comment by gclough** [comment]
Similar to #90 , and likely the same cause.

**Comment by nikolaev-rd** [comment]
Not sure, but maybe this issue relates to #134

**Comment by kenanrizvic** [comment]
> We have the same problem. Another problem, that we have: it seems like equal user from database used password from `auth_file` on PgBouncer, not from PostgreSQL database.

That sentence helped me to find problem in my case. We had in database section, lines with both user=XY and auth_user=XY  and password in userlist.txt  for XY user  and it worked ok untill someone try to login with unexisting user where pgbouncer failed with:

> 2019-09-03 13:49:03.155 29836 FATAL @src/objects.c:312 in function put_in_order(): put_in_order: found existing elem

Solution was to remove auth_user from database section.

**Comment by petere** [comment]
The problem is that in PgBouncer internally, users with the same name can exist multiple times in different contexts:
- globally (from `auth_file`)
- PAM user
- attached to a database (via `auth_query`)
- as a "forced user" (`user=` entry in a database)

This is in principle correct, since for example users obtained via `auth_query` from different databases could have different passwords.

At the same time, however, PgBouncer maintains a global list of pools (global variable `pool_list`) that it thinks is uniquely identifiable by database name and user name. You can already see how this would be a problem.

In the particular case, where you have

    mydb = ... user=foo auth_user=foo

it creates a global user `foo` for the `auth_user=` parameter and a forced user `foo` for the `user=` parameter. When it then launches the `auth_query`, it looks for a pool with db=mydb/user=foo, which initially won't exist, so it creates it and adds it to the global pool list. It also adds it to the pool list associated with the (global) `foo` user. Then, later, when authentication is done, it switches to the (forced) user `foo` and looks for a pool with db=mydb/user=foo in that user's pool list. But that user hasn't been used yet, so it creates a new pool with db=mydb/user=foo. That all works fine, but then it will also add that new pool to the global pool list, which will fail with the "found existing elem" error.

I think most of the variously reported `put_in_order()` crashes are due to this or a similar confusion between global and non-global users.

One could alleviate this specific instance by putting a check into `force_user()` (which creates the forced user) that checks if the user already exists globally. However, this would not work if the order of actions is the other way around, such as when the forced user is created first and then a global user is added later (for example, through a configuration reload). Also, this kind of configuration should be allowed, I think, so erroring out might be the wrong choice.

Another possibility might be to just remove the "found existing elem" check and allow duplicate user/db combinations in the global pool list. In a quick try, this appears to work, but this would clearly need further analysis to check that it's safe.

**Comment by taybin** [comment]
@ndavidson-bigpay I'm seeing this as well.

**Comment by tyteen4a03** [comment]
Hi, any updates to this?

**Comment by wwoytenko** [comment]
Hi!
I have found the source of the problem and just made  [PR](https://github.com/pgbouncer/pgbouncer/pull/817). Have a look at it, I have described the issue.

It throws because we have admin/stats user defined in `pgbouncer.ini`, but it does not exist either in `userlist.txt` or in the PostgreSQL server

Anyway, a few recommendations which should fix this problem with current builds
* Turn on `verbose = 2` in order to understand which connection the pgbouncer falls apart. In my case, I have found `(nodb)/(nouser)@[::1]:33194 got var: user=stats`
* Check that `admin_users` or `stats_users` exist in the DB in which we are performing `auth_query`. 
* If it does not exist we may do any of these actions:
  * Remove this user from `admin_users` or `stats_users`. 
  * Add user's creds into `userlist.txt`
  * Add user in the PostgreSQL server in which proxying pgbouncer
* Additionally I recommend to hardcode `auth_dbname` in order to prevent using the admin database (such as pgbouncer) as an auth database

**Comment by JelteF** [comment]
Reopening this as #817 didn't fix the problem after all.

**Comment by drekthral** [comment]
> > We have the same problem. Another problem, that we have: it seems like equal user from database used password from `auth_file` on PgBouncer, not from PostgreSQL database.
> 
> That sentence helped me to find problem in my case. We had in database section, lines with both user=XY and auth_user=XY and password in userlist.txt for XY user and it worked ok untill someone try to login with unexisting user where pgbouncer failed with:
> 
> > 2019-09-03 13:49:03.155 29836 FATAL @src/objects.c:312 in function put_in_order(): put_in_order: found existing elem
> 
> Solution was to remove auth_user from database section.

After many headaches this solution worked for us.

Pgbouncer crashed every time it received request with user that was not defined in pgbouncer. Resulting in container restart. With this vulnerability everyone could keep us down with just sending request to our pgbouncer with random name and password.

**Comment by JelteF** [comment]
@drekthral yeah it's definitely not a fun bug, a PR to fix is is definitely appreciated.

**Comment by phemmer** [comment]
I might also suggest adjusting [pgbouncer.service](https://github.com/pgbouncer/pgbouncer/blob/master/etc/pgbouncer.service) with a `Restart=always`, or at the very least `Restart=on-failure`. Because otherwise, yeah, it's really easy to cause downtime, especially when admin has to manually bring it back up. I've worked around the issue on my systems by creating a drop-in and overriding it, but should probably be standard.

**Comment by JelteF** [comment]
> I was able to reproduce this locally with both 1.21 and 1.20. The important bit is to set the `user` in the database line:
> mydb = host=localhost **user=anybody** port=5432 dbname=mydb
> 
> PR #817 does not include tests for that case.

Adding my quote from the other issue here too. To keep the description of the issue in one place.

**Comment by JelteF** [comment]
Closing this since @eulerto reported that this was fixed at least in 1.17 already.
